Concurrent modification exception when building

A `ConcurrentModificationException` may occur when building.

> I tried to do this:

Build a project.

> I expected this to happen:

The build to succeed.

> Instead, this happened:

It failled with this exception:

```
2016-12-23 13:25:29,781 [  35522]  ERROR - .builders.MetaborgMetaBuilder2 - An unexpected exception occurred.
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1461)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadComponentRange(DefaultLanguageManager.java:179)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadImplRange(DefaultLanguageManager.java:190)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadRange(DefaultLanguageManager.java:200)
	at org.metaborg.intellij.jps.builders.JpsSpoofaxMetaBuilder.afterBuild(JpsSpoofaxMetaBuilder.java:186)
	at org.metaborg.intellij.jps.builders.SpoofaxPostBuilder.doBuild(SpoofaxPostBuilder.java:80)
	at org.metaborg.intellij.jps.builders.SpoofaxPostBuilder.doBuild(SpoofaxPostBuilder.java:34)
	at org.metaborg.intellij.jps.builders.MetaborgMetaBuilder2.build(MetaborgMetaBuilder2.java:93)
	at org.metaborg.intellij.jps.builders.MetaborgMetaBuilder2.build(MetaborgMetaBuilder2.java:40)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTarget(IncProjectBuilder.java:979)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:960)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1021)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:908)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:734)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:387)
	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:195)
	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:138)
	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:295)
	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:125)
	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler$1.run(BuildMain.java:238)
	at org.jetbrains.jps.service.impl.SharedThreadPoolImpl$1.run(SharedThreadPoolImpl.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
2016-12-23 13:25:29,798 [  35539]   INFO - .incremental.IncProjectBuilder - java.util.ConcurrentModificationException
org.jetbrains.jps.incremental.ProjectBuildException: java.util.ConcurrentModificationException
	at org.metaborg.intellij.jps.builders.MetaborgMetaBuilder2.build(MetaborgMetaBuilder2.java:103)
	at org.metaborg.intellij.jps.builders.MetaborgMetaBuilder2.build(MetaborgMetaBuilder2.java:40)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTarget(IncProjectBuilder.java:979)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:960)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1021)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:908)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:734)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:387)
	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:195)
	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:138)
	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:295)
	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:125)
	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler$1.run(BuildMain.java:238)
	at org.jetbrains.jps.service.impl.SharedThreadPoolImpl$1.run(SharedThreadPoolImpl.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1461)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadComponentRange(DefaultLanguageManager.java:179)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadImplRange(DefaultLanguageManager.java:190)
	at org.metaborg.intellij.languages.DefaultLanguageManager.unloadRange(DefaultLanguageManager.java:200)
	at org.metaborg.intellij.jps.builders.JpsSpoofaxMetaBuilder.afterBuild(JpsSpoofaxMetaBuilder.java:186)
	at org.metaborg.intellij.jps.builders.SpoofaxPostBuilder.doBuild(SpoofaxPostBuilder.java:80)
	at org.metaborg.intellij.jps.builders.SpoofaxPostBuilder.doBuild(SpoofaxPostBuilder.java:34)
	at org.metaborg.intellij.jps.builders.MetaborgMetaBuilder2.build(MetaborgMetaBuilder2.java:93)
	... 18 more
```

---

Spoofax 2.1.0 release

--- a/org.metaborg.spoofax-common/src/main/java/org/metaborg/intellij/languages/DefaultLanguageManager.java
+++ b/org.metaborg.spoofax-common/src/main/java/org/metaborg/intellij/languages/DefaultLanguageManager.java
@@ -176,7 +176,7 @@ public class DefaultLanguageManager implements ILanguageManager {
      */
     @Override
     public void unloadComponentRange(final Iterable<ILanguageComponent> components) {
-        for (final ILanguageComponent component : components) {
+        for (final ILanguageComponent component : Lists.newArrayList(components)) {
             unload(component);
         }
     }
@@ -186,7 +186,7 @@ public class DefaultLanguageManager implements ILanguageManager {
      */
     @Override
     public void unloadImplRange(final Iterable<? extends ILanguageImpl> impls) {
-        for (final ILanguageImpl impl : impls) {
+        for (final ILanguageImpl impl : Lists.newArrayList(impls)) {
             unloadComponentRange(impl.components());
         }
     }
@@ -196,7 +196,7 @@ public class DefaultLanguageManager implements ILanguageManager {
      */
     @Override
     public void unloadRange(final Iterable<ILanguage> languages) {
-        for (final ILanguage language : languages) {
+        for (final ILanguage language : Lists.newArrayList(languages)) {
             unloadImplRange(language.impls());
         }
     }

